pkg/vere/mdns.c

```diff
@@ -0,0 +1,220 @@
+  snprintf(who2, sizeof who2, "~%s", who);
+  payload->who = who2;
+
+  struct addrinfo hints;
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_INET; // Request only IPv4 addresses
+  hints.ai_socktype = SOCK_STREAM; // TCP socket
+
+  uv_getaddrinfo_t* req = (uv_getaddrinfo_t*)c3_malloc(sizeof(uv_getaddrinfo_t));
+  req->data = (void*)payload;
+
+  uv_loop_t* loop = uv_default_loop();
+  int error = uv_getaddrinfo(loop, req, getaddrinfo_cb, host, NULL, &hints);
+
+  if (error < 0) {
+    u3l_log("mdns: getaddrinfo error: %s\n", uv_strerror(error));
+    c3_free(payload);
+    c3_free(req);
+    DNSServiceRefDeallocate(sref);
+  }
+}
+
+static void poll_cb(uv_poll_t *handle, int status, int events) {
+  DNSServiceRef sref = (DNSServiceRef) handle->data;
+  int err = DNSServiceProcessResult(sref);
+}
+
+static void dns_sd_cb(DNSServiceRef sref, mdns_payload* payload) {
+  int fd = DNSServiceRefSockFD(sref);
+  uv_loop_t* loop = uv_default_loop();
+  uv_poll_t* poll = (uv_poll_t *)c3_malloc(sizeof(uv_poll_t));
+  poll->data = (void *)sref;
+  payload->poll = poll;
+  uv_poll_init(loop, poll, fd);
+  uv_poll_start(poll, UV_READABLE, poll_cb);
+}
+
+static void browse_cb(DNSServiceRef s,
+                      DNSServiceFlags f,
+                      uint32_t interface,
+                      DNSServiceErrorType err,
+                      const char* name,
+                      const char* type,
+                      const char* domain,
+                      void* context)
+{
+  if (err != kDNSServiceErr_NoError) {
+    u3l_log("mdns: service browse error %i", err);
+    return;
+  }
+
+  if (f & kDNSServiceFlagsAdd) {	// Add
+    DNSServiceRef sr;
+
+    mdns_payload* payload = (mdns_payload*)context;
+    mdns_payload* payload_copy = c3_malloc(sizeof *payload_copy);
+
+    // copy to prevent asynchronous thrashing of payload
+    memcpy(payload_copy, payload, sizeof(mdns_payload));
+
+    DNSServiceErrorType err =
+      DNSServiceResolve(&sr, 0, interface,
+                        name, type, domain, resolve_cb,
+                        (void*)payload_copy);
+
+    if (err != kDNSServiceErr_NoError) {
+      u3l_log("mdns: dns service resolve error %i", err);
+      c3_free(payload_copy);
+      DNSServiceRefDeallocate(sr);
+      return;
+    }
+    dns_sd_cb(sr, payload_copy);
+  }
+}
+
+static void register_cb(DNSServiceRef sref,
+                        DNSServiceFlags f,
+                        DNSServiceErrorType err,
+                        const char* name,
+                        const char* type,
+                        const char* domain,
+                        void* context)
+{
+  mdns_payload* payload = (mdns_payload*)context;
+
+  if (err != kDNSServiceErr_NoError) {
+    u3l_log("mdns: service register error %i", err);
+  } else {
+    u3l_log("mdns: %s registered on all interfaces", name);
+  }
+  uv_poll_stop(payload->poll);
+  c3_free(payload->poll);
+  c3_free(payload->our);
+  c3_free(payload);
```

Comment:
leave a comment explaining why we leak `sref`.

Reply:
Done